### PR TITLE
gRPC API proposal (draft)

### PR DESCRIPTION
**Pull request by arbimo**
Here is PR to discuss the protobuf/gRPC API. A previous proposal was done in #66 together with a prototype implementation.

This PR only contains the `.proto` file because:
 - it defines a public interface and should be discussed (almost) independently from the internals
 - it would facilitate discussion and reviewing to have a single file in the PR

The previous proposal had several limitations. Some were just easy to fix missing parts, some were more fundamental: the model would work for classical planning but scaling it to richer models would be very awkward.

### Content

Everything (including comments and discussion) is in the attached `upf.proto` file.

In a nutshell, I propose a new protobuf format that tries to model in a unified way planning problem with different features.
The design space is pretty large and I tried to present two representative way of modeling a planning problem in many cases. 

Anytime two solutions are proposed, my personal preference is with the second one, however other might disagree with that and this thread should be the occasion of sorting this out. Since this is a PR, you can directly comment on the code or even push you own changes.

### Future work

- [x] Decide between the various proposals in the file
- [x] Complete the encoding of temporal planning
- [ ] Document the association of fields in some message with some problem features (e.g. the `temporal_qualification` field can only be set if the `durative_actions` feature is set). This is essential for:
    - limiting the complexity for consumers ("I can safely ignore this field since I do not support this feature")
    - allowing for backward compatibility when extending the protobuf format

### Links

[Protobuf documentation](https://developers.google.com/protocol-buffers/docs/proto3)

### CODE REVIEW

**Comment by arbimo** [comment]
Tagging interesting parties: @mikand @alvalentini @uwe-koeckemann @franklinselva

**Comment by arbimo** [comment]
I just pushed an update of the proposal. Following the conclusions of this morning discussion, it is a mostly complete and coherent model (just missing some implementation details for durative actions).

This confirms that having a unique action representation is the most appealing way.
Essentially, the model is designed in such a way that a consumer can act as if the field it does not support did not exist.
In the model, a field that is annotated with `features: [X, Y]`, it can simply be ignored by any planner not supporting features `X` and `Y`.
In my opinion this is very appealing to:

 - simplify the overall consumption of action models for planners that support several features (at no cost for those that don't)
 - avoid the combinatorial explosion of action kinds
 - allow backward compatible extension of planning problems
 - live up to the "unified planning" name :partying_face: 




## Expression encoding

The one point for which it was unclear how we should go this morning was related to expression representation.
In the previous proposal from Uwe (which closely mirrors the implementation of the UP), there was an explicit enumerated field that was used to find out whether a node was:
 - an atom
 - a fluent name associated with some arguments
 - an application of the `+` function to some arguments
 - an application of the `-` function to some arguments
 - ...

This had the advantage of greatly facilitating parsing, but it forces us to be exhaustive in the functions that we support. Do we support the `cos` function ? the `sqrt`, `log2`, `tanh` function ? How would we support a user specified function (e.g. in the context of semantic attachment).
 
For the sake of flexibility/extensibility, an option could be to support, in addition to the built-in functions, some user specified functions (whose name would a string). I see two problems with this:

 - there would be two ways of having function application (one for built-in and one for user-defined/planner-specific)
 - we could not add new built-in functions without breaking backward compatibility: a planner using a user-defined "cos" would break once this is moved to a built-in `COS` by the upf. 

The current `.proto` proposal moves away from this while attempting to retain the ease of parsing. In essence, it works as an s-expression: an expression is either an atom or a list of expressions. See the `Expression` message in the `.proto` file.


The ease of parsing is maintained by having an expression be annotated with its kind which allows testing the shape of the expression. In addition, the type of the expressions are provided which would allow distinguishing the rare overloaded functions.
It should simply be completed with the string representation of built-in functions (e.g. "=" but not "eq" for equality). To avoid naming collision when adding built-in, it should be absolutely non ambiguous what the function means (this is the case for most mathematical functions).

In essence, the new proposal thus relies on string-based function names but with added meta-data to facilitate consumption.
I currently sees this as the proposal with the least number of downsides but I am not entirely convinced this is the best possible design. Also it puts the responsibility of correctly annotating the expressions on the UP.


As an exercise, below is what it takes to convert these string-based functions into my own set of supported functions or to test whether an expression is an addition (in rust, other languages might be more verbose ;).
Here the test on `kind` is essential to be able to test a particular field without a crash/null-pointer-exception.


```rust
/// Functions supported in my planner
enum Function {
    Add,
    Sub,
    Cos,
    ShortestPath // semantic attachment
}

/// Attempts to transform an expression into a function that we support.
fn as_function(expr: &Expression) -> Result<Function> {
    if expr.kind == ExpressionKind::FUNCTION_SYMBOL {
        match expr.atom.symbol {
            "+" => Ok(Function::Add),
            "-" => Ok(Function::Sub),
            "cos" => Ok(Function::Cos),
            "shortest_path" => Ok(Function::ShortestPath)
            _ => Err("unsupported function"),
        }
    } else {
        Err("Not a function")
    }
}

/// tests whether this is application of the function `function_name`
fn is_application_of(expr: &Expression, function_name: &str) -> bool {
    expr.kind == ExpressionKind::FUNCTION_APPLICATION 
      && expr.list[0].atom.symbol == function_name
}
// is this an addition such as `(+ a 1 2)`
fn is_addition(expr: &Expression) -> bool {
    is_application_of(expr, "+")
}

```

**Comment by madmage** [comment]
I like very much the discussion here, but I think the topic is much larger and deeper than gRPC.
In my opinion, the gRPC API should be exactly the same we have for the UP itself (I mean the Python interface, and possibly the gRPC should be somehow autogenerated to avoid inconsistencies).

Moreover, I guess we should have a different discussion (in the Issue tab here) for each topic: what kind of actions will the UP support? what kind of problems? what kind of partial response can give the planner? etc.

**Comment by arbimo** [comment]
To clarify my intention with this PR: the objective is not to set the interface and associated features in stone but rather to have a baseline that we can further refine.
The proposed `.proto` is large to make sure that the design is globally coherent and extensible. The downside being that it cover to many things to easily discuss them.
My suggested course of action would be to:
 - merge this PR once the overall structure of the model is considered good enough
 - discuss improvements on particular aspects in separate PR that update this baseline. For instance, the discussion on `Expression` or on the partial response would be much easier to discuss as a diff on this baseline as they are mostly independent from the other aspects.

**Comment by alvalentini** [comment]
In order to agree on a common terminology, I list the differences between the proto file and the UP:

| Proto File | UP | Notes |
|:-------------:|:----:|:--------:|
| RelativeTime | Timing | |
| Timepoint + Extractor | bool is_from_start | field of RelativeTime/Timing <br> no distinctions between global/not global in the UP |
| delay | bound | field of RelativeTime/Timing |
| TemporalQualification | Interval | in the proto file there are no distinctions between open or closed interval |
| interval_start | lower | field of TemporalQualification/Interval |
| interval_end | upper | field of TemporalQualification/Interval |
| Duration | DurationInterval | |
| lower_bound | lower | field of Duration/DurationInterval |
| upper_bound | upper | field of Duration/DurationInterval |
| strict_lower_bound | is_left_open | field of Duration/DurationInterval |
| strict_upper_bound | is_right_open | field of Duration/DurationInterval |
| TimedFact | TimedEffect | |
| operator | kind | field of the effect class |

**Comment by arbimo** [comment]
@alvalentini: I just push a update following most of the naming suggestions you mentionned.I updated your answer, putting in bold the names I retained.


I created the `Interval` type that contains two `Expression`s and fields to distinguish between open/closed interval, following the field names you indicated.

The `Duration` now contains a single `Interval` field called `controllable_in_bounds`.
This is in line with your `DurationInterval` class that should be seen as a specialization of a `Duration` class. Other specialization of `Duration` should be possible in the future (such as the `FixedDuration` mentionned in temporal planning discussion).

I kept the `TemporalQualification` message distinct from `Interval` as it its fields are of different type (it refers to `Timing` and not `Expression`).
Maybe we should specifically discuss what is meant by the qualification and whether it makes sense to split the qualification between one for conditions (essentially an interval) and one for effect (a single timepoint?).


Regarding the `Timing` structure, I don't really understand the idea behind the `bound` name in the UP and find the `delay` more meaningful (`delta` could be another candidate).

I also think it makes sense to keep the `Timepoint` message separated with a specific field in `Timing`. 
In the future it is very likely that we will have to refer to timepoints beyond the start/end of the container. I have in mind at least the timepoints related to the start/end of subtasks (for hierarchical planning), to the start/end of an arbitrary action in the plan (for execution or richer plan representation), ...


Regarding the presence of `GLOBAL_START`/`GLOBAL_END` it is useful to avoid subtle bugs.
My understanding of the UP is that the interval `[50,70]` would be expressed `[START+50, START+70]` which is only correct if the interval is then used in a context where `START` actually refers to the global start.

**Comment by mikand** [comment]
@arbimo Thanks for this pass!

@alvalentini can you contribute to this PR by:
- align the Timing class with protobuf proposal
- Rename `Interval` -> `TemporalQualification`
- Align start/end fields of intervals/temporal qualification

Concerning the `Duration` vs `DurationInteval`, what about calling it `DurationConstraint`? In Python we can create a small class hierarchy for duration constraints, in protobuf, just like for actions, we create a single message

@arbimo I think that left/right openness specification is missing in `TemporalQualifier` in the proto, no?

**Comment by alvalentini** [comment]
@arbimo I rebased the branch on top of master

**Comment by arbimo** [comment]
> Concerning the `Duration` vs `DurationInteval`, what about calling it `DurationConstraint`? In Python we can create a small class hierarchy for duration constraints, in protobuf, just like for actions, we create a single message

Note that if the duration in non controllable, then it is not a constraint which was the motivation behind using the `Duration` name (can be seen as short for duration-specification). That being said, I am also fine with `DurationConstraint` as it is inline with what many will expect.

> @arbimo I think that left/right openness specification is missing in `TemporalQualifier` in the proto, no?

That's right. I did not include as I am not sure how the described interval should relate to a qualified effect.
What are the patterns currently supported by the UP for effects ? Depending on the answer, it might make sense to have a `TemporalInterval` as the qualifier for conditions and a single `Timing` as the qualifier of the effect. (or distinguishing between them in another way.

**Comment by mikand** [comment]
@arbimo 

- OK for using `Duration` instead of `DurationConstraint`.

- ATM the UP only allows effects on specific times (i.e. a single `TIming`). If we keep this limitation, I think that having Effects to be associated with a `Timing` is better to avoid confusion. If, instead, we want to generalize to the case with effects over intervals (a la ANML `:->`) then having intervals on both conditions and effects seems better. Shall we discuss this tomorrow at the WP3 meeting?

**Comment by arbimo** [comment]
As a baseline I pushed a commit splitting the `TemporalQualification` into a `TimeInterval` for conditions and a `Timing` for effects.
For conditions, I think it very natural.

For effects, the timepoint should represent the time at which the effect is fulfilled. I assume that there will always be such a time regardless of the effect kind. I can be completed with an additional (optional) timepoint to represent the "start of the effect process", which might be useful for the ANML-like semantics of `[s,e] sv == v1 :-> v2;` or numeric effects (resource consumption).

Caveat: the limitation of having two `Timing` instead of a `TimeInterval` is that we can't express the "left/right open" (i.e. immediately before/after).

**Review by alvalentini — COMMENTED** [review]
@mikand I updated the code following the renaming we discussed.

@arbimo I added some minor comments to the proto file.

File: unified_planning/grpc/unified_planning.proto [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+message Effect {
+    // Required. The actual effect that should take place.
+    EffectExpression effect = 1;
+    // Optional. If the effect is conditional, then the following field must be set.
+    // In this case, the `effect` will only be applied if the `condition`` holds.
+    // TODO: When should the condition hold in case of durative actions? Is this supported by PDDL? Two options:
+    //        - immediately before the effect (generalization of the non-temporal case)
+    //        - have the condition field be of type `Condition` which allows for temporal qualification. If not qualified, the above would apply.
+    // features: conditional_effects
+    Expression condition = 2;
```
[the remark this answers is not in the document]
Comment (arbimo) [reply unified_planning/grpc/unified_planning.proto]: I implemented the change.
I preferred the previous encoding: having a basic `EffectExpression` that is annotated with possible `condition` and `occurrence_time` in the `Effect` message (i.e. all qualifications are in the same place).
I am not going to fight over this so if you feel it is better this way that's fine with me ;)
If you change your mind, you can just revert this change.

Note that the documentation of the condition does not need to be that complex now that a single `Timing` is used.

File: unified_planning/grpc/unified_planning.proto [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+// ============== Problem =========================
+
+
+// A Goal is currently an expression that must hold in the final state.
+// It is in its own message to allow its extension with (e.g.):
+//  - temporal qualification
+//  - weight (for optional goals, oversubscription planning)
+message Goal {
+    // Goal expression that must hold in the final state.
+    Expression goal = 1;
```
[the remark this answers is not in the document]
Comment (arbimo) [reply unified_planning/grpc/unified_planning.proto:268]: Great, didn't know it was supported already !

**Comment by mikand** [comment]
@arbimo For us it would be important to merge this ASAP as the renaming has an impact on other PRs.

My proposal is to go ahead and merge this and start to maintain the proto file. Then we will address the three points you mentioned one-by-ne in separate PRs?

Sounds good?

@uwe-koeckemann any feedback?

**Comment by arbimo** [comment]
@mikand Sounds good to me.

**Pull request merged**